### PR TITLE
[CHIA-1553] Replace pylint with ruff

### PR DESCRIPTION
**Pull request by Quexington**
Blatant plagiarism of https://github.com/Chia-Network/chia-blockchain/pull/18649 with less emojis and politics :)

Ruff is a python linter and formatter written entirely in Rust that aims to reimplement many of the most popular tools in a way where they can run hundreds of times faster.  It's used to lint the source of many major projects like Mypy, Poetry, PyInstaller, pytest, and even Pylint, which this PR hopes to replace.

The introduced scope of ruff here is very limited compared to what it can do.  We're _only replacing pylint_ in this PR even though this tool could in theory absorb 4 or 5 more of our pre-commit jobs (and run them all in a fraction of the time).  The reason for targeting pylint is two-fold: 1) pylint cannot run in pre-commit because it's so slow.  This often results in unexpected CI failures that devs would have much rather caught at commit time.  2) pylint at once wants to check type related information and at the same time is pretty bad at it.  This is a known issue in pylint that results in many false positives and while they're working to improve it, they are slow to respond to false positive issues and have seeming architectural issues that prevent them from addressing some of the more annoying failures like with respect to wrappers and dataclasses.

Ruff does not have exact 1:1 parity with pylint, however, they are working in that direction: https://github.com/astral-sh/ruff/issues/970.  They also don't attempt type inference that they're bad at, which means we're trading some linting coverage for less false positives and `pylint: disable=` statements in our codebase.

In terms of Ruff ergonomics, many errors can be fixed automatically (and are by pre-commit) and errors that cannot be fixed give much more descriptive formatting to help developers solve them.  Ignoring ruff errors works like flake8 (because it can replace it) by saying `# noqa: <error code>`.

### CODE REVIEW

**Comment by emlowe** [comment]
@SocketSecurity ignore pypi/ruff@0.7.1

**Comment by wallentx** [comment]
Pff.. chicken

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -106,3 +106,8 @@ repos:
         entry: ./activated.py flake8
         language: system
         types: [python]
+  - repo: https://github.com/astral-sh/ruff-pre-commit
+    rev: v0.7.1
+    hooks:
+      - id: ruff
+        args: ["--fix"]
```
Comment (altendky) [inline .pre-commit-config.yaml:113]: I'm at least tempted to run it out of our local venv like other tools here.  The primary upside is a single definition of the version we use in a form that all our existing dependencies tools are monitoring.

File: pyproject.toml [hunk 2]
```diff
@@ -6,9 +6,9 @@ description = "Chia blockchain full node, farmer, timelord, and wallet."
 authors = ["Mariano Sorgente <mariano@chia.net>"]
 license = "Apache License"
 readme = "README.md"
-keywords= ["chia", "blockchain", "node"]
+keywords = ["chia", "blockchain", "node"]
```
Comment (altendky) [inline pyproject.toml:9]: is the reformatting throught this file intentionally being introduced by ruff?  or...  prettier is supposed to be processing toml so i'm a bit annoyed that this is was apparently so messy.

File: pyproject.toml [hunk 3]
```diff
@@ -108,10 +110,34 @@ pyupgrade = { version = "3.16.0", optional = true }
 # numpy = [
 # 	{version="1.24.4", python = "<3.9", optional = true},
 # 	{version="1.26.4", python = ">=3.9", optional = true}]
+ruff = "^0.7.1"
```
Comment (altendky) [inline pyproject.toml:113]: shall we har dpin like everything else for consistency?  pending some bugs github needs to address, we plan on changing this pattern for many dependencies.

```suggestion
ruff = "0.7.1"
```

File: chia/wallet/wallet_protocol.py [hunk 4]
```diff
@@ -60,7 +60,7 @@ async def match_hinted_coin(self, coin: Coin, hint: bytes32) -> bool: ...
     wallet_info: WalletInfo
     # WalletStateManager is only imported for type hinting thus leaving pylint
     # unable to process this
```
Comment (altendky) [inline chia/wallet/wallet_protocol.py:62]:
```suggestion
```

File: chia/wallet/util/debug_spend_bundle.py [hunk 5]
```diff
@@ -12,7 +12,7 @@
 from chia.util.hash import std_hash
 from chia.wallet.uncurried_puzzle import UncurriedPuzzle
 
-CONDITIONS = {k: bytes(v)[0] for k, v in ConditionOpcode.__members__.items()}  # pylint: disable=E1101
+CONDITIONS = {k: bytes(v)[0] for k, v in ConditionOpcode.__members__.items()}
```
Comment (altendky) [inline chia/wallet/util/debug_spend_bundle.py:15]: i think this stuff that was being complained about was also not needed.  but, i'll explore that in another pr.  just fyi that a real smell was being pointed out here.

https://github.com/Chia-Network/chia-blockchain/pull/18767

**Review by altendky — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml (hunk 1 above)
Comment (Quexington) [reply .pre-commit-config.yaml]: Good idea, I didn't think about that.

File: pyproject.toml (hunk 3 above)
Comment (Quexington) [reply pyproject.toml]: Taken

File: chia/wallet/wallet_protocol.py (hunk 4 above)
Comment (Quexington) [reply chia/wallet/wallet_protocol.py]: Deleted

File: pyproject.toml (hunk 2 above)
Comment (Quexington) [reply pyproject.toml:9]: You know, I'm not sure.  There were a few stages of getting this PR ready locally.  One iteration ran the formatter and it's possible out of a desire to keep the changes to this file, I accidentally kept all of the formatting as well.  That's my best guess, but I can confirm if you like.  The formatting seems good as you noted, so I don't see a harm in keeping it around.

Comment (Quexington) [reply pyproject.toml:9]: Confirmed that this does not get re-added if I remove some of it.  But like I said, I think we should keep it.

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -106,3 +106,10 @@ repos:
         entry: ./activated.py flake8
         language: system
         types: [python]
+  - repo: local
+    hooks:
+      - id: ruff
+        name: Ruff
+        entry: ./activated.py ruff check --fix
+        language: system
+        pass_filenames: false
```
Comment (altendky) [inline .pre-commit-config.yaml:115]: can we still let pre-commit handle the filenames?  i guess ruff is faster so maybe we can just inefficiently check everything, but...  `:]`

File: pyproject.toml (hunk 2 above)
Comment (altendky) [reply pyproject.toml:9]: i think we're just missing a prettier plugin for handling toml files.  i'm working on adding that.  would you be willing to narrow this down to just the 'real' changes and i'll swing back through afterwards with a pr where prettier cleans it up?  assuming i really get it working.

**Review by altendky — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml (hunk 6 above)
Comment (Quexington) [reply .pre-commit-config.yaml]: done

**Review by altendky — APPROVED** [review]

**Review by emlowe — APPROVED** [review]

**Review by cmmarslender — APPROVED** [review]

**Comment by emlowe** [comment]
Coverage diff exemption

**Pull request merged**